python/src/specialize.cc

```diff
@@ -0,0 +1,884 @@
+    } else {
+      type_str = u64_str;
+      if (specialize_value) {
+        key = (align && ((val_64 & 15) == 0)) ? D_str : empty_str;
+      }
+    }
+  }
+
+  return {py::handle(type_str), py::handle(key)};
+}
+
+static std::pair<py::handle, py::handle>
+handle_tensor(PyObject *backend, PyObject *arg, bool is_const,
+              bool specialize_value, bool align) {
+  // no DECREF on these as they are used for objects we cache or return them
+  PyObject *canon_res = nullptr, *type_str = nullptr, *key_res = nullptr;
+  // normal temporary objects
+  PyObject *dtype = nullptr, *native_spec_obj = nullptr;
+  PyObject *data_ptr_method = nullptr, *data_ptr_result = nullptr;
+  PyObject *fallback_method = nullptr, *kwargs = nullptr, *args_tuple = nullptr;
+
+  py::handle key = nullptr;
+  Py_hash_t dtype_hash = -1;
+  DTypePtrKey dsk{0, false};
+  DtypePtr2Str::iterator it;
+  bool native_impl_available = false;
+  uint64_t data_ptr = 0;
+
+  // handle type_str specialization of a tensor
+  dtype = PyObject_GetAttr(arg, dtype_attr);
+  if (!dtype)
+    goto cleanup;
+
+  dtype_hash = PyObject_Hash(dtype);
+  if (dtype_hash == -1)
+    goto cleanup;
+
+  dsk = {dtype_hash, is_const};
+  it = dtype_ptr2str.find(dsk);
+
+  if (it != dtype_ptr2str.end()) {
+    type_str = it->second;
+  } else {
+    canon_res =
+        PyObject_CallFunctionObjArgs(canonicalize_ptr_dtype_fn, dtype,
+                                     is_const ? Py_True : Py_False, nullptr);
+    if (!canon_res)
+      goto cleanup;
+    dtype_ptr2str[dsk] = canon_res;
+    type_str = canon_res;
+  }
+
+  // handle alignment specialization of a tensor
+  if (!specialize_value) {
+    key = py::handle();
+    Py_DECREF(dtype);
+    return {py::handle(type_str), key};
+  }
+
+  native_spec_obj = PyObject_GetAttr(backend, has_native_tensor_spec_attr);
+  if (native_spec_obj) {
+    native_impl_available = PyObject_IsTrue(native_spec_obj);
+  } else {
+    PyErr_Clear();
+    // on error we fall back to native_impl_available = false gracefully
+  }
+
+  if (native_impl_available) {
+    data_ptr_method = PyObject_GetAttr(arg, data_ptr_attr);
+    data_ptr_result = PyObject_CallNoArgs(data_ptr_method);
+    if (!data_ptr_result)
+      goto cleanup;
+
+    data_ptr = PyLong_AsUnsignedLongLong(data_ptr_result);
+    if (PyErr_Occurred())
+      goto cleanup;
+
+    key = (align && ((data_ptr & 15) == 0)) ? D_str : empty_str;
+  } else {
+    // call fallback
+    fallback_method = PyObject_GetAttr(backend, get_tensor_spec_attr);
+    if (!fallback_method)
+      goto cleanup;
+
+    kwargs = PyDict_New();
+    args_tuple = PyTuple_Pack(1, arg);
+    if (kwargs && args_tuple) {
+      if (align) {
+        PyDict_SetItemString(kwargs, "align", Py_True);
+      }
+      key_res = PyObject_Call(fallback_method, args_tuple, kwargs);
+      if (key_res) {
+        key = py::handle(key_res);
+      } else {
+        goto cleanup;
+      }
+    } else {
+      goto cleanup;
+    }
```

Comment:
```suggestion
    PyObject *args[3] = {backend, arg, align ? Py_True : Py_False};
    PyObject *kwnames = align_tuple;
    key_res = PyObject_VectorcallMethod(get_tensor_spec_attr, args, 3, kwnames);
    if (!key_res)  goto cleanup;
    key = py::handle(key_res);
```

(where `align_tuple` is a global equal to `("align",)`

Reply:
Also key is owned here, so should be wrapped in `py::object` rather than `py::handle`. I think in general the refcounting is wrong for the results from these functions as some branches return new references while others return borrowed references.

Reply:
I agree that most of these things are not ideal. The outermost function returns a tuple which should call `INCREF` on the types and keys returned by these helper functions. Cached objects should be fine - as the cache owns them and the reference count should stay at 1, and simply forwarding `arg` (e.g. for constexpr types or for the integer 1) also should be fine. Function calls used to compute auxiliary values or to fill the cache also should be fine. 

IIUC, this case of `key` then should be wrong as I am missing a DECREF after the tuple creation. Would having
```
key = py::reinterpret_steal<py::object>(key_res)
```
be fine then?

Reply:
Yes, I would change these helpers to return `pair<py::object, py::object>` so the ownership is clearer.

Reply:
This is slightly slower but in a similar range as the caching. Gained robustness likely justifies this switch.